lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -1395,19 +1396,27 @@ bool swift::extractCompilerFlagsFromInterface(
   // we have loaded a Swift interface from a different-but-compatible
   // architecture slice. Use the compatible subarchitecture.
   for (unsigned I = 1; I < SubArgs.size(); ++I) {
-    // FIXME: Also fix up -target-variant (rdar://135322077).
-    if (strcmp(SubArgs[I - 1], "-target") != 0) {
-      continue;
+    if (strcmp(SubArgs[I - 1], "-target") == 0) {
+      llvm::Triple target(SubArgs[I]);
+      if (PreferredTarget &&
+          target.getSubArch() != PreferredTarget->getSubArch() &&
+          target.getArch() == PreferredTarget->getArch() &&
+          target.getVendor() == PreferredTarget->getVendor() &&
+          target.getOS() == PreferredTarget->getOS() &&
+          target.getEnvironment() == PreferredTarget->getEnvironment())
+        target.setArch(PreferredTarget->getArch(), PreferredTarget->getSubArch());
+      SubArgs[I] = ArgSaver.save(target.str()).data();
+    } else if (strcmp(SubArgs[I - 1], "-target-variant") == 0) {
+      llvm::Triple targetVariant(SubArgs[I]);
+      if (PreferredTargetVariant &&
+          targetVariant.getSubArch() != PreferredTargetVariant->getSubArch() &&
```

Comment:
Could this comparison be factored out into a helper function/lambda to avoid repeating the conditions?